Added PublishCondition feature to the ReportEntryExtension.

Michael1993 opened the pull request:
Hopefully a bit better in quality than my last PR. Progresses #179 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

aepfli commented:
> Kudos, SonarCloud Quality Gate passed!
> 
> [<img alt="Bug" width="16" height="16" src="https://camo.githubusercontent.com/67545b394988d492007fcd881c0a94c5296edc29/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=BUG) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/9701f681ab497016e6ae7f63fbaf44f4ab30d34c/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=BUG)
> [<img alt="Vulnerability" width="16" height="16" src="https://camo.githubusercontent.com/475ba60a579e3b8ba1e2a251f4a264312f56d9b3/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=VULNERABILITY) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/9701f681ab497016e6ae7f63fbaf44f4ab30d34c/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=VULNERABILITY) (and [<img alt="Security Hotspot" width="16" height="16" src="https://camo.githubusercontent.com/cdf79f6d9d06866802c83878aa3be0bd576eb7f9/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=SECURITY_HOTSPOT) to review)
> [<img alt="Code Smell" width="16" height="16" src="https://camo.githubusercontent.com/8747c0c3ceeb057fd2942c9cef31347d3bcbb6d4/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=CODE_SMELL) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/9701f681ab497016e6ae7f63fbaf44f4ab30d34c/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=CODE_SMELL) [12 Code Smells](https://sonarcloud.io/project/issues?id=junit-pioneer_junit-pioneer&pullRequest=216&resolved=false&types=CODE_SMELL)
> 
> [<img alt="100.0%" width="16" height="16" src="https://camo.githubusercontent.com/f6ad00bbcd688dc63ac44f8127929a2fc92c58e6/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f3130302e706e67">](https://sonarcloud.io/component_measures?id=junit-pioneer_junit-pioneer&pullRequest=216&metric=new_coverage&view=list) [100.0% Coverage](https://sonarcloud.io/component_measures?id=junit-pioneer_junit-pioneer&pullRequest=216&metric=new_coverage&view=list)
> [<img alt="0.0%" width="16" height="16" src="https://camo.githubusercontent.com/540735ca9aeb74d34ef0758d1adaeeb0d1c173e7/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332e706e67">](https://sonarcloud.io/component_measures?id=junit-pioneer_junit-pioneer&pullRequest=216&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=junit-pioneer_junit-pioneer&pullRequest=216&metric=new_duplicated_lines_density&view=list)

okay most of the code smells are not useful, should we deactivate this rule regarding assertions?

Bukama commented on src/main/java/org/junitpioneer/jupiter/ReportEntry.java:
```diff
@@ -45,4 +45,36 @@
+	/**
+	 * This makes the {@code ReportEntry} repeatable.
+	 */
+	@Retention(RetentionPolicy.RUNTIME)
```
Note: I appreciate the declartion of the repeatable annotation inside it (as I suggested it in #215 ). But as the discussion on that has not really started I can not say if this is fine or has to be changed later.

Michael1993 replied:
I see no reason to store this is a separate file. You can not use it by itself, only in combination with the `@ReportEntry`.

aepfli commented on src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:
```diff
@@ -13,20 +13,28 @@
 import static java.util.stream.Collectors.toList;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junitpioneer.jupiter.ReportEntry.PublishCondition.*;
```
because noboby currently runs the checkstyle target for test, we only run checkstyleMain (like it used to be) but we would need to run also checkstyleTest (or something like that, not 100% sure about the naming

aepfli replied:
```
Task :checkstyleTest FAILED
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junit/jupiter/engine/AbstractJupiterTestEngineTests.java:0: error: Missing package-info.java file.
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junit/platform/engine/test/TestDescriptorStub.java:0: error: Missing package-info.java file.
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junit/platform/engine/test/event/ExecutionEvent.java:0: error: Missing package-info.java file.
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junitpioneer/AbstractPioneerTestEngineTests.java:0: error: Missing package-info.java file.
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junitpioneer/jupiter/DefaultLocaleTests.java:0: error: Missing package-info.java file.
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junitpioneer/jupiter/params/RangeSourceProviderTests.java:0: error: Missing package-info.java file.
[ant:checkstyle] /home/sschrottner/projects/github/junit-pioneer/src/test/java/org/junitpioneer/vintage/TestIntegrationTests.java:0: error: Missing package-info.java file.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkstyleTest'.
> Checkstyle rule violations were found. See the report at: file:///home/sschrottner/projects/github/junit-pioneer/build/reports/checkstyle/test.html
  Checkstyle files with violations: 7
  Checkstyle violations by severity: [error:7]
```

Bukama replied:
Quoting @nicolaiparlog here:

> [14:18] nipafx: At first glance it seems that most of those should apply to tests as well, assuming there aren't any defaults hidden somewhere. We should discuss that on the issue.

Bukama commented on src/main/java/org/junitpioneer/jupiter/ReportEntry.java:
```diff
@@ -45,4 +45,36 @@
+	 *  - ALWAYS:		publish report entry every time the test is ran, regardless of its outcome
+	 *  - ON_SUCCESS:	publish report entry after successful test run
+	 *  - ON_FAILURE:	publish report entry after failed test run
+	 *  - NEVERMORE:	never publishes report entry
```
Invalid value :)

Bukama commented on src/main/java/org/junitpioneer/jupiter/ReportEntryExtension.java:
```diff
@@ -33,4 +42,31 @@ private static void verifyKeyValueAreNotBlank(ReportEntry entry) {
+	private void publishOnConditions(ExtensionContext context, ReportEntry.PublishCondition... conditions) {
+		findAnnotations(context)
+				.filter(entry -> Arrays.asList(conditions).contains(entry.when()))
+				.filter(entry -> !entry.key().isEmpty() && !entry.value().isEmpty())
```
Is this line necessary as non-empty keys and values are checked in line 31 ?

Michael1993 replied:
It's necessary, if you want to avoid throwing an exception from TestReporter.

nicolaiparlog replied:
I don't get it either. As far as I understand the code, this line will only ever do something if `verifyKeyValueAreNotBlank` is broken by a change. But such a change would fail a test and thus not go undetected. I hence don't see a scenario where this line changes anything.

Michael1993 replied:
We throw an `ExtensionConfigurationException` before test execution if either key or value are blank. This leads to the test failing. So, JUnit calls the `TestWatcher::testFailed`. If the publish condition is `ALWAYS` or `ON_FAILURE`, we try to publish report entries. So this is filtering out *why* the test failed - we don't want to publish if the extension configuration is incorrect.

Michael1993 replied:
Added a short comment that (tries to) concisely explains this.

Bukama commented on src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:
```diff
@@ -69,20 +82,158 @@ void emptyValue_fails() {
 	}
 
+	@Nested
+	@DisplayName("with explicitly set when parameter")
```
maybe put the `when` into quotes or so. As of now it reads a bit rough (at least to me)

nicolaiparlog commented:
I wondered whether `PublishCondition.ON_DISABLED` would make sense. On the surface, it doesn't, but maybe someone needs it. Then again, as was pointed out to me, such messages should probably go into the `@Disabled` annotation, which have a `value` attribute for exactly that. Another aspect I didn't come up with myself: `@ReportEntry`-'s selling mantra is "`TestReporter` as an annotation" and if the test is disabled, you couldn't programmatically call the `TestReporter`.

Makes sense, let's not have PublishCondition.

nicolaiparlog commented:
One more, maybe `PublishCondition.ON_ABORTED`? Seems to be an interesting use case, to provide additional report entries if tests were aborted. Question is, do aborted tests also count as failed?

I think they shouldn't, so `ON_ABORTED` and `ON_FAILURE` are mutually exclusive. Their Javadoc should point that out, though.

Bukama commented:
> One more, maybe `PublishCondition.ON_ABORTED`? Seems to be an interesting use case, to provide additional report entries if tests were aborted. Question is, do aborted tests also count as failed?
> 
> I think they shouldn't, so `ON_ABORTED` and `ON_FAILURE` are mutually exclusive. Their Javadoc should point that out, though.

I don't see the use case that I will annotate all my message for the maybe case of an abortion. It would duplicate all other (failure) annotations. Personally I don't care if a test fails because of an exception or a failed assertion - it fails!

nicolaiparlog commented:
In "Nest new test cases" (a75d7bb), I placed the new tests into more nested classes so it's more obvious which combinations we tested. By the looks of it, we're missing some combinations. Should they be added?

![screenshot_20200414_200938](https://user-images.githubusercontent.com/6537432/79258849-0c978800-7e8c-11ea-966a-25762eff3590.png)

nicolaiparlog commented:
> Personally I don't care if a test fails because of an exception or a failed assertion - it fails!

A failed assertion leads to a failed test, not an aborted test. Tests get aborted by assumptions and... nothing else that I can come up with right now. My `@RepeatFailedTest` reports initial failed test runs as aborted, but I'm not sure whether it does it in a way that this extension can pick up.

nicolaiparlog commented:
Just tried it and `@RepeatFailedTest` apparently does things right and failed-but-marked-as-disabled tests could emit specific messages. I find that a quite interesting cooperation between these two features.

Bukama commented:
I dig a bit deeper into assumptions. They shall be used if you want to check the preconditions for your tests. As test should be independent I only see them useful in distributed OS related tests, like selenium tests, where you can't use tags like `@EnableOnLinux` and have to check them inside the test preparation. Another use case I can imagine are integration tests where you want to check the environment is ready before executing your tests. As I sadly don't have much experience with such tests I can't say if my assumption (haha) is meaningfull.

But I agree that there may be cases where `ON_ABORTED` can make sense. And as it's easy to implement, we should offer it too.

Michael1993 commented:
Nicolai made a very appealing case for `ON_ABORTED`, so I added it. I also added the missing(?) test cases so the test looks much nicer (even if a bit redundant). I added the missing 's' to `ReportEntryExtensionTests`.

I added a comment that hopefully clears up why filtering empty keys/values is necessary before publishing.

A lot more of "The Raven" got into the PR (Yay! :🎉)

nicolaiparlog commented:
After the accidental merge, I opened #247 - the Pull Request Of Shame.
